org.eclipse.xtext/src/org/eclipse/xtext/validation/NamesAreUniqueValidationHelper.java

```diff
@@ -24,62 +21,152 @@
 import org.eclipse.emf.ecore.EcorePackage;
 import org.eclipse.xtext.naming.QualifiedName;
 import org.eclipse.xtext.resource.IEObjectDescription;
+import org.eclipse.xtext.resource.ISelectable;
 import org.eclipse.xtext.service.OperationCanceledManager;
+import org.eclipse.xtext.util.CancelIndicator;
 
+import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Maps;
+import com.google.inject.Singleton;
 
 /**
  * @author Sebastian Zarnekow - Initial contribution and API
  */
+@Singleton
 public class NamesAreUniqueValidationHelper implements INamesAreUniqueValidationHelper {
 
 	private ImmutableSet<EClass> clusterTypes = getClusterTypes();
-	
-	@Inject 
+
+	@Inject
 	private OperationCanceledManager operationCanceledManager = new OperationCanceledManager();
-	
+
 	/**
-	 * <p>Initialize the set of clustering types. A type is considered to be clustering
-	 * if any instance of that type has to have a unique name when
-	 * it is transformed to an {@link org.eclipse.xtext.resource.IEObjectDescription}.
-	 * Instances that do not belong to the same cluster may have the same exported name.</p>
-	 * <p>A clustering type will often be some kind of root type in a type hierarchy.</p> 
+	 * <p>
+	 * Initialize the set of clustering types. A type is considered to be clustering if any instance of that type has to
+	 * have a unique name when it is transformed to an {@link org.eclipse.xtext.resource.IEObjectDescription}. Instances
+	 * that do not belong to the same cluster may have the same exported name.
+	 * </p>
+	 * <p>
+	 * A clustering type will often be some kind of root type in a type hierarchy.
+	 * </p>
 	 */
 	protected ImmutableSet<EClass> getClusterTypes() {
 		return ImmutableSet.of(EcorePackage.Literals.EOBJECT);
 	}
-	
+
+	/**
+	 * @deprecated Implementations should adhere to the context provided via
+	 *             {@link #checkUniqueNames(Iterable, Context, ValidationMessageAcceptor)} since 2.22
```

Comment:
maybe we can place `since 2.22` at the beginning of the sentence to better emphasize that. At the end it is very difficult to recognize. Occurs on other places as well.